### PR TITLE
feat: add TTS Live Client

### PR DESCRIPTION
**Pull request by SandraRodgers**
Introduce live speak functionality to existing `speak` client. Maintain existing interface for batch.

## Summary by CodeRabbit

- **New Features**
	- Added detailed documentation for using the Deepgram Text-to-Speech API, including REST and WebSocket methods.
	- Introduced live text-to-speech functionality, allowing real-time audio synthesis.
	- Added new classes for managing text-to-speech requests and live connections, enhancing user interaction.

- **Bug Fixes**
	- Simplified export structure for the SpeakRestClient class.

- **Documentation**
	- Updated README with practical examples for both synchronous and asynchronous TTS implementations.
	- Expanded documentation with event handling details for live text-to-speech synthesis.

### CODE REVIEW

File: src/packages/AbstractLiveClient.ts [hunk 1]
```diff
@@ -258,6 +259,48 @@ export abstract class AbstractLiveClient extends AbstractClient {
...
+    // To be implemented by subclasses
+  }
+
+  /**
+   * Handles binary messages received from the WebSocket connection.
+   * @param data - The binary data.
+   */
+  protected handleBinaryMessage(data: ArrayBuffer): void {
+    // To be implemented by subclasses
+  }
```
Comment (lukeocodes) [inline src/packages/AbstractLiveClient.ts:303]: this appears to be incomplete? what are subclasses?

**Review by lukeocodes — CHANGES_REQUESTED** [review]

File: examples/node-speak-live/index.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+const { createClient, LiveTTSEvents } = require("../../dist/main/index");
+const fetch = require("cross-fetch");
+
+const live = async () => {
+  const text = "Hello, how can I help you today?";
+
+  const deepgram = createClient(process.env.DEEPGRAM_API_KEY, {
+    global: { fetch: { options: { url: "https://api.beta.deepgram.com" } } },
```
Comment (naomi-lgbt) [inline examples/node-speak-live/index.js:8]:
```suggestion
    global: { websocket: { options: { url: "wss://api.beta.deepgram.com" } } },
```

It seems to look for a websocket override, not a fetch override.

File: examples/node-speak-live/index.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  });
+
+  const connection = deepgram.speak.live({ text }, { model: "aura-asteria-en" });
```
Comment (naomi-lgbt) [inline examples/node-speak-live/index.js:11]:
```suggestion
  const connection = deepgram.speak.live({ model: "aura-asteria-en" });
```

The function definition for this expects `SpeakSchema, endpoint` - passing the `model` object as the second argument broke the function.

File: src/packages/AbstractLiveClient.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@ import { AbstractClient, noop } from "./AbstractClient";
 import { CONNECTION_STATE, SOCKET_STATES } from "../lib/constants";
 import type { DeepgramClientOptions, LiveSchema } from "../lib/types";
```
Comment (lukeocodes) [inline src/packages/AbstractLiveClient.ts:3]: You shouldn't need to make any changes to the Abstract client for this work. These edits should instead live inside `SpeakLiveClient.ts`

File: src/packages/index.ts [hunk 5]
```diff
@@ -7,4 +7,4 @@ export * from "./ListenRestClient";
 export * from "./ManageRestClient";
 export * from "./ReadRestClient";
 export * from "./SelfHostedRestClient";
-export * from "./SpeakRestClient";
```
Comment (lukeocodes) [inline src/packages/index.ts:10]: this file is called a barrel file. it should export all clients that exist in the directory. please export `SpeakRestClient`, `SpeakLiveClient`, and `SpeakClient`

File: src/packages/SpeakClient.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * 2. `live(ttsOptions: SpeakSchema = {}, endpoint = ":version/speak")`: Returns a `SpeakLiveClient` instance for interacting with the live speak API, with the provided TTS options and endpoint.
+ */
+export class SpeakClient extends AbstractClient {
+  public namespace: string = "speak";
+
+  /**
+   * Returns a `SpeakRestClient` instance for interacting with the rest speak API.
+   */
+  get request() {
+    return new SpeakRestClient(this.options);
```
Comment (lukeocodes) [inline src/packages/SpeakClient.ts:20]: This is good, I assume to provide backwards compatibility with the dot notation we set up for TTS REST

Comment (lukeocodes) [reply src/packages/SpeakClient.ts:20]: however I think for this one, there is a slight gotcha with how we initially set it up, and how to maintain the compatibility of it.

If you make `get request(){}` look like this...

```
  get request(
    source: TextSource,
    options?: SpeakSchema,
    endpoint = ":version/speak"
  ) {
    const client = new SpeakRestClient(this.options);
 
    return client.request(source, options, endpoint);
  }
```

it _should_ maintain the interface that exists now to do `deepgram.speak.request(...etc)`, despite adding a class between so we can also do `deepgram.speak.live(...etc)`

**Review by lukeocodes — CHANGES_REQUESTED** [review]

File: src/packages/index.ts (hunk 5 above)
Comment (lukeocodes) [reply src/packages/index.ts:10]: For this, I use a tool called auto-barrel which allows you to right-click on a directory and click on "update barrel file", which does it all for you.

File: README.md [hunk 7]
```diff
@@ -382,10 +382,18 @@ const vttOutput = webvtt(result);
 
 # Text to Speech
 
+## Rest
+
 ```js
 const { result } = await deepgram.speak.request({ text }, { model: "aura-asteria-en" });
 ```
 
+## Websocket
```
Comment (lukeocodes) [inline README.md:391]: Can you check out the other websocket example. It's a little more comprehensive than just the connection.

**Review by lukeocodes — APPROVED** [review]
Do we want to release this on alpha/beta or straight to main while the API is still in development?

**Comment by jpvajda** [comment]
@lukeocodes  Things we need to probably change in this PR:

* Rename of reset to clear
* Clear message now has a response type.
* We want to remove the container field 

cc @dvonthenen

**Comment by dvonthenen** [comment]
> @lukeocodes Things we need to change in this PR:
> 
> * Rename of reset to clear
> * Clear message now has a response type.
> * We want to remove the container field
> 
> cc @dvonthenen

Yep, confirmed those should be the only changes between now and towards the end of July. 👍

**Review by jpvajda — APPROVED** [review]
LGMT!

**Pull request merged**